Python/import.c

```diff
@@ -303,10 +303,120 @@ _PyImport_Fini(void)
 PyObject *
 PyImport_GetModuleDict(void)
 {
-    PyInterpreterState *interp = PyThreadState_GET()->interp;
-    if (interp->modules == NULL)
-        Py_FatalError("PyImport_GetModuleDict: no module dictionary!");
-    return interp->modules;
+    PyObject *sysdict = PyThreadState_GET()->interp->sysdict;
+    if (sysdict == NULL)
+        Py_FatalError("PyImport_GetModuleDict: no sys module!");
+
+    _Py_IDENTIFIER(modules);
+    PyObject *modules = _PyDict_GetItemId(sysdict, &PyId_modules);
+    if (modules == NULL)
+        Py_FatalError("lost sys.modules");
+    return modules;
+}
+
+/* In some corner cases it is important to be sure that the import
+   machinery has been initialized (or not cleaned up yet).  For
+   example, see issue #4236 and PyModule_Create2(). */
+
+int
+_PyImport_IsInitialized(PyInterpreterState *interp)
+{
+    if (interp->sysdict == NULL)
+        return 0;
+    _Py_IDENTIFIER(modules);
+    PyObject *modules = _PyDict_GetItemId(interp->sysdict, &PyId_modules);
+    if (modules == NULL)
+        return 0;
+    return 1;
+}
+
+PyObject *
+_PyImport_GetModule(PyObject *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
+    if (PyDict_Check(modules))
```

Comment:
I would prefer to test PyDict_CheckExact().

Reply:
Done.